torch/nn/_functions/rnn.py

```diff
@@ -10,26 +10,29 @@
 
 
 def RNNReLUCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.relu(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+    x_h = input if w_ih is None else F.linear(input, w_ih, b_ih)
+    hy = F.relu(x_h + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
 def RNNTanhCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
-    hy = F.tanh(F.linear(input, w_ih, b_ih) + F.linear(hidden, w_hh, b_hh))
+    x_h = input if w_ih is None else F.linear(input, w_ih, b_ih)
+    hy = F.tanh(x_h + F.linear(hidden, w_hh, b_hh))
     return hy
 
 
 def LSTMCell(input, hidden, w_ih, w_hh, b_ih=None, b_hh=None):
     if input.is_cuda:
-        igates = F.linear(input, w_ih)
+        igates = input.expand(4, input.size(0), input.size(1)).transpose(0, 1) if w_ih is None else F.linear(input,
```

Comment:
Just to make clear, the 'bug' is that the input to all 4 gate-wise calculations will have the same weight input matrix (Wi,Wf,Wo,Wc) are all the same? And what you think should happen is the input should be an input of dimension hidden * 4 (separate for each gate?)

If this is the case I'm not seeing what is being broken by this implementation other than it reflecting the cuDNN behaviour (which can be argued wrong).